coral-schema/src/main/java/com/linkedin/coral/schema/avro/MergeHiveSchemaWithAvro.java

```diff
@@ -0,0 +1,262 @@
+ * 6. For fields found only in Hive schema, field names are sanitized to make them compatible with Avro identifier spec
+ */
+class MergeHiveSchemaWithAvro extends HiveSchemaWithPartnerVisitor<Schema, Schema.Field, Schema, Schema.Field> {
+
+  static Schema visit(StructTypeInfo typeInfo, Schema schema) {
+    return HiveSchemaWithPartnerVisitor.visit(typeInfo, schema, new MergeHiveSchemaWithAvro(),
+        AvroPartnerAccessor.INSTANCE);
+  }
+  private final AtomicInteger recordCounter = new AtomicInteger(0);
+
+  @Override
+  public Schema struct(StructTypeInfo struct, Schema partner, List<Schema.Field> fieldResults) {
+    boolean shouldResultBeOptional = partner == null || isNullableType(partner);
+    Schema result;
+    if (partner == null || extractIfOption(partner).getType() != Schema.Type.RECORD) {
+      // if there was no matching Avro struct, return a struct with new record/namespace
+      int recordNum = recordCounter.incrementAndGet();
+      result = Schema.createRecord("record" + recordNum, null, "namespace" + recordNum, false);
+      result.setFields(fieldResults);
+    } else {
+      result = SchemaUtilities.copyRecord(extractIfOption(partner), fieldResults);
+    }
+    return shouldResultBeOptional ? SchemaUtilities.makeNullable(result) : result;
+  }
+
+  @Override
+  public Schema.Field field(String name, TypeInfo field, Schema.Field partner, Schema fieldResult) {
+    // No need to infer `shouldResultBeOptional`. We expect other visitor methods to return optional schemas
+    // in their field results if required
+    if (partner == null) {
+      // if there was no matching Avro field, use name form the Hive schema and set a null default
+      return new Schema.Field(SchemaUtilities.makeCompatibleName(name), fieldResult, null, null);
+    } else {
+      // TODO: How to ensure that field default value is compatible with new field type generated from Hive?
+      // Copy field type from the visitor result, copy everything else from the partner
+      // Avro requires the default value to match the first type in the option, reorder option if required
+      Schema reordered = reorderOptionIfRequired(fieldResult, partner.defaultValue());
+      return SchemaUtilities.copyField(partner, reordered);
+    }
+  }
+
+  /**
+   * Reorders an option schema so that the type of the provided default value is the first type in the option schema
+   *
+   * e.g. If the schema is (NULL, INT) and the default value is 1, the returned schema is (INT, NULL)
+   * If the schema is not an option schema or if there is no default value, schema is returned as-is
+   */
+  private Schema reorderOptionIfRequired(Schema schema, Object defaultValue) {
+    if (isNullableType(schema) && defaultValue != null) {
+      boolean isNullFirstOption = schema.getTypes().get(0).getType() == Schema.Type.NULL;
+      // TODO: verify if null works
+      if (isNullFirstOption && defaultValue.equals(null)) {
```

Comment:
Condition `defaultValue.equals(null)` is always `false` when reached. I think we can change type of defaultValue to be `JsonNode` rather than `Object` and call `defaultValue.isNull()`